Add VM support

monstermunchkin opened the pull request.

stgraber commented on image/lxd.go:
```diff
@@ -37,12 +37,17 @@ func NewLXDImage(sourceDir, targetDir, cacheDir string,
 
+	if vm {
+		// If we're building a VM, we need the separate metadata tarball.
+		unified = false
```
Technically unified is possible with a VM, it would be a single tarball containing the metadata files and templates and a `rootfs.img` file (qcow2) instead of the `rootfs` directory.

stgraber commented on distrobuilder/scripts.go:
```diff
@@ -0,0 +1,199 @@
+    dd if=/dev/zero of="$1" bs=1 count=0 seek="${IMAGESIZE}"
+}
+
+create_partitions() {
```
So I think we'll want all our images to look like:
 - GPT partition table only
 - First partition 100MB vfat EF00 (/boot/efi)
 - Second partition is / with configurable filesystem and size

A bunch of the other bits in those scripts feel like they should be either file generators or be scripts in the VM definition. I would expect distrobuilder to take care of creating the image and partitions, formatting them and mounting them where appropriate. It should then have a generator that gets us a basic fstab based on the partitions `blkid` and should expose the block device paths of the EFI and root partitions through environment so the scripts can handle it.

tomponline replied:
@stgraber just a thought here, would it be possible to label the partition in such a way that we could detect when initialising a VM instance and resize the correct filesystem after unpacking it onto its block device/file so that its the same size as the volume? The ubuntu images do this with cloud-init I believe, but would be good if this works for non-cloudinit users too.

stgraber replied:
Yeah, I think the rootfs label is what's used for that. We'll definitely want to make sure that our cloud-init enabled images do the right thing, for the rest, we could add a basic systemd unit in distrobuilder which would achieve the same, at least on ext4.

stgraber commented on distrobuilder/main.go:
```diff
@@ -230,6 +230,90 @@ func (c *cmdGlobal) preRunBuild(cmd *cobra.Command, args []string) error {
+			return err
+		}
+
+		partUUID, err := getRootfsPartitionUUID()
```
To generate the fstab, we'll actually need the `blkid` for both the UEFI and root partitions.

stgraber commented on distrobuilder/vm.go:
```diff
@@ -0,0 +1,146 @@
+
+	var stdout bytes.Buffer
+
+	err := lxd.RunCommandWithFds(nil, &stdout, "kpartx", "-a", "-s", "-v", imageFile)
```
Can you see if you can't avoid `kpartx` altogether?
It's pretty annoying to run inside containers, so if you can instead use `losetup` with a calculated offset, that would be much more compatible with running distrobuilder in jenkins.

stgraber commented on distrobuilder/main_lxd.go:
```diff
@@ -41,6 +42,7 @@ func (c *cmdLXD) commandBuild() *cobra.Command {
 	c.cmdBuild.Flags().StringVar(&c.flagType, "type", "split", "Type of tarball to create"+"``")
 	c.cmdBuild.Flags().StringVar(&c.flagCompression, "compression", "xz", "Type of compression to use"+"``")
 	c.cmdBuild.Flags().BoolVar(&c.flagVM, "vm", false, "Create a qcow2 image for VMs"+"``")
+	c.cmdBuild.Flags().UintVar(&c.flagImageSize, "size", 2147483648, "Size of the VM image"+"``")
```
Any reason why we wouldn't have this in the yaml instead?
Could be a key under targets/lxd

stgraber commented:
We're going to want a file generator that can add the systemd units for the agent.

stgraber commented:
```
root@v1:~/lxc-ci/images# /root/go/bin/distrobuilder build-dir ubuntu.yaml /tmp/out -o image.release=focal
Error: flag accessed but not defined: vm
```

stgraber commented:
```
Error: Error while downloading source: symlink gutsy /usr/share/debootstrap/scripts/focal: no such file or directory
```

When missing debootstrap, a better error would be nice.

stgraber commented:
```
Error: Failed to create LXD image: exec: "qemu-img": executable file not found in $PATH
```

After 10min of building :)
Would be good to check for the dependencies ahead of doing stuff.

stgraber commented:
Hmm, that's going to be a bit of a problem:
```
-rw-r--r--  1 root root 1.0T Feb 10 23:20 ubuntu-focal-amd64-default-20200210_2319.raw
```

Can we default to 5GB instead so it actually fits in LXD? :)

stgraber commented:
http://paste.ubuntu.com/p/xTYbnvBWTx/ for the size thing

stgraber commented:
So unlike Ubuntu that needs some pretty complex booting logic, we really don't need all that complexity for LXD and so don't need to play too many games with the partition table.

A simple GPT table with two partitions should do the trick:
 - 1: EF00 vfat 250MB
 - 2: 8300 ext4 2GB (or whatever fs the user decides to use)

Then we load up shim and grub onto the EFI partition, point it to the main partition using its GPT UUID and we should be good to go. I'll see if I can't figure out the changes for that.

stgraber commented:
http://paste.ubuntu.com/p/VkDyQt2c2d/ for the partition order

stgraber commented:
@monstermunchkin can you provide the YAML you're using for testing?
I was looking at the EFI setup, but given my YAML doesn't install the usual suspects, it's a bit empty :)

stgraber commented:
I'm confused by the `fstab` generator, why does the `name` property have to be the filesystem name when we have a separate property for this already in YAML which the generator already consumes?

stgraber commented:
```
stgraber@castiana:~/vm$ sudo cat /mnt/etc/fstab 
LABEL=rootfs  /         ext4  defaults  0 0
	LABEL=UEFI    /boot/efi vfat  defaults  0 0
	stgraber@castiana:~/vm$ 
```

That also doesn't look quite right.

stgraber commented:
http://paste.ubuntu.com/p/bcxvs27TYx/ to bump to 4GB and avoid duplication

stgraber commented:
http://paste.ubuntu.com/p/HVwKNW7Yg9/ is my current YAML.
This gets me a working grub though there's something wrong with both the menu entries (extra content) and secureboot.

monstermunchkin commented:
I fixed most of the issues you mentioned (systemd generator is still missing).

> I'm confused by the fstab generator, why does the name property have to be the filesystem name when we have a separate property for this already in YAML which the generator already consumes?

I updated the docs and the fstab generator itself regarding the `name` property. It's not needed.

My config was [this one](https://paste.ubuntu.com/p/GQ4Dm4JY3f/), with code copied from the `livecd-rootfs` package.

monstermunchkin commented:
This is what I get when checking the console:
```
BdsDxe: loading Boot0005 "ubuntu" from HD(1,GPT,9EF6BEE2-3F74-4F47-A0EF-2033FAD35CE0,0x800,0x32000)/\EFI\ubuntu\shimx64.efi
BdsDxe: starting Boot0005 "ubuntu" from HD(1,GPT,9EF6BEE2-3F74-4F47-A0EF-2033FAD35CE0,0x800,0x32000)/\EFI\ubuntu\shimx64.efi
error: no suitable video mode found.
EFI stub: UEFI Secure Boot is enabled.
```

stgraber commented:
Yeah, same here, this suggests that we're not booting properly through shim or are otherwise missing some signed kernel bits.

stgraber commented:
Left to do:
 - Add a `types` condition which can take `container` or `vm` and defaults to both and support it wherever we accept `architectures`, `releases` and `variants`.
 - Add a `lxd-agent` generator that takes `init` as a configuration key and when set to `systemd` spits out the usual set of units and symlinks.

stgraber commented:
We also need to make it so that we can `build-dir` to get a rootfs and then `pack-lxd` with or without the VM flag and get the expected artifacts.

I think the best approach here is to make use of overlayfs to setup a writable overlay on top of the output directory of build-dir. In that overlay we can then install and run operations that are specific to the target without having to care about reverting those changes.

monstermunchkin commented:
I cleaned up the commits a bit, and added the `types` filter and the overlayfs support which works well for VMs.
Things left to do:
- [ ] lxd-agent generator
- [ ] Ensure things still work properly for containers

stgraber commented:
There's also missing logic when running split `build-dir` and `pack-lxd --vm`.
I did:
 - /root/go/bin/distrobuilder build-dir images/ubuntu.yaml /root/rootfs
 - /root/go/bin/distrobuilder pack-lxd images/ubuntu.yaml /root/rootfs /root/img/ --vm

With: http://paste.ubuntu.com/p/j2qHyFfMmV/

This failed with:

```
root@djanet:~/lxc-ci# /root/go/bin/distrobuilder pack-lxd images/ubuntu.yaml /root/rootfs /root/img/ --vm
Creating new GPT entries.
Warning: The kernel is still using the old partition table.
The new table will be used at the next reboot or after you
run partprobe(8) or kpartx(8)
GPT data structures destroyed! You may now partition the disk using fdisk or
other utilities.
Creating new GPT entries.
Warning: The kernel is still using the old partition table.
The new table will be used at the next reboot or after you
run partprobe(8) or kpartx(8)
The operation has completed successfully.
Warning: The kernel is still using the old partition table.
The new table will be used at the next reboot or after you
run partprobe(8) or kpartx(8)
The operation has completed successfully.
mke2fs 1.44.1 (24-Mar-2018)
Discarding device blocks: done                            
Creating filesystem with 1022715 4k blocks and 511488 inodes
Filesystem UUID: 34268bbf-be14-4afe-b551-77f02a8ee6ed
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376, 294912, 819200, 884736

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (16384 blocks): done
Writing superblocks and filesystem accounting information: done 

mkfs.fat 4.1 (2017-01-24)
+ passwd -u root
passwd: password expiry information changed.
+ echo root:root
+ chpasswd
+ update-grub
sh: 7: update-grub: not found
Error: Failed to run post-files: exit status 127
```

That's because anything that was restricted to the `vm` type was ignored as far as I can tell.

Entries that have a `type` restriction need to be applied during packing rather than building which likely means we need some extra logic there (unless I missed it already in this branch).

monstermunchkin commented:
You're having that problem because I decided to make the `--vm` flag global again and the flag therefore applies to `build-dir` as well. This made packaging and post actions easier to deal with. I'll add some extra logic to fix that issue.

monstermunchkin commented:
@stgraber Now it should work properly. The `--vm` flag is for LXD only, and I added some extra logic for the VM type.

stgraber commented:
Anything we're still missing?

stgraber commented:
@monstermunchkin can you split the build/pack reshuffle into its own branch, then also make the lxd-agent generator its own branch?

That should then make this branch only contain the new bits for VMs, making things a bit easier and faster to land.

monstermunchkin commented:
@stgraber I split up this PR into [VM preparations](#277) and [lxd-agent generator](#276). Once both are merged, I'll update this PR which should then contain only 3 commits.

monstermunchkin commented:
Rebased and ready for review.
